modules/images/webp-uploads/load.php

```diff
@@ -35,5 +35,186 @@ function webp_uploads_filter_image_editor_output_format( $output_format, $filena
+ * metadata of the attachment, on the other hand a `sources` property is added.
+ *
+ * @since n.e.x.t
+ *
+ * @see   wp_generate_attachment_metadata
+ * @see   webp_uploads_valid_image_mime_types
+ *
+ * @param array $metadata      An array with the metadata from this attachment.
+ * @param int   $attachment_id The ID of the attachment where the hook was dispatched.
+ *
+ * @return array An array with the updated structure for the metadata before is stored in the database.
+ */
+function webp_uploads_create_sources_property( array $metadata, $attachment_id ) {
+	// This should take place only on the JPEG image.
+	$valid_mime_types = webp_uploads_valid_image_mime_types();
+
+	// Not a supported mime type to create the sources property.
+	if ( ! array_key_exists( get_post_mime_type( $attachment_id ), $valid_mime_types ) ) {
+		return $metadata;
+	}
+
+	// All subsizes are created out of the original image.
+	$file = wp_get_original_image_path( $attachment_id, true );
+
+	// File does not exist.
+	if ( ! file_exists( $file ) ) {
+		return $metadata;
+	}
+
+	// Prevent to convert JPEG to WebP if we are creating JPEG versions of the image.
+	remove_filter( 'image_editor_output_format', 'webp_uploads_filter_image_editor_output_format' );
+
+	$sizes = array();
+	foreach ( webp_uploads_get_image_sizes() as $size => $properties ) {
+		$image_sizes = array();
+		if ( array_key_exists( 'sizes', $metadata ) && is_array( $metadata['sizes'] ) ) {
+			$image_sizes = $metadata['sizes'];
+		}
+
+		$current_size = array();
+		if ( array_key_exists( $size, $image_sizes ) && is_array( $image_sizes[ $size ] ) ) {
+			$current_size = $image_sizes[ $size ];
+		}
+
+		$sources = array();
+		if ( array_key_exists( 'sources', $current_size ) && is_array( $current_size['sources'] ) ) {
+			$sources = $current_size['sources'];
+		}
+
+		// Try to find the mime type of the image size.
+		if ( array_key_exists( 'mime-type', $current_size ) ) {
+			$current_mime = $current_size['mime-type'];
+		} elseif ( array_key_exists( 'file', $current_size ) ) {
+			$current_mime = wp_check_filetype( $current_size['file'] )['type'];
+		} else {
+			$current_mime = '';
+		}
+
+		// The mime for this file couldn't be determined.
+		if ( empty( $current_mime ) ) {
+			continue;
+		}
+
+		// Make sure the current mime is consider a valid mime type.
+		if ( ! array_key_exists( $current_mime, $valid_mime_types ) ) {
+			continue;
+		}
+
+		$sources[ $current_mime ] = array(
+			'file' => array_key_exists( 'file', $current_size ) ? $current_size['file'] : '',
+			// TOOD: Add filesize from the original version of this image.
+		);
+
+		$formats = array_diff_assoc( $valid_mime_types, array( $current_mime => $valid_mime_types[ $current_mime ] ) );
+
+		foreach ( $formats as $mime => $extension ) {
+			// Editor needs to be recreated every time as there is not flush() or clear() function that can be used after we created an image.
```

Comment:
Similar to my above comment, we should only do any thing here if this image size does not yet exist in this MIME type. Otherwise we'll endlessly run duplicate processes to create image sizes. This is critical because the `wp_generate_attachment_metadata` filter is also invoked when missing image sizes are being created (see also WordPress core's `wp_get_missing_image_subsizes()` function), so we have to also only create the image sizes here that _do not_ yet exist in the format.

Basically add this:
```suggestion
			if ( isset( $sources[ $mime ] ) ) {
				continue;
			}

			// Editor needs to be recreated every time as there is not flush() or clear() function that can be used after we created an image.
```

Reply:
Agree, There's a third parameter that indicates if this comes from `create` or `update` called context that we can use to optimize this condition for those use cases.